### PR TITLE
Fixes dissonance between `disconnect_expired_cert` vs `require_session_mfa`

### LINKED ISSUE #18544: `disconnect_expired_cert: yes` and `require_session_mfa: yes` can't harmoniously exist on one cluster
If an auth service has both `disconnect_expired_cert` and `require_session_mfa` enabled, users are consistently kicked out of their sessions after about a minute (across features). The issue is that with per-session mfa we typically issue 1 minute-lived certs per session, which `disconnect_expired_cert` then identifies as expired and boots the user out.

The proposed solution is that some metadata can be added to the MFA cert that contains the original TTL of the user cert that made the request. New sessions using the MFA cert can still be denied based on the MFA cert TTL, but existing sessions can be closed based on the original TTL of the user cert when `disconnected_expired_certs` is enabled and the original TTL of the user cert is present. (credit to @dboslee for this proposal).

### PR DESCRIPTION
**Pull request by ibeckermayer**
Adds ssh/tls cert extensions
```
// CertExtensionMFAVerifiedSessionExpires is the RFC3339 timestamp representing the hard
// deadline of the session on a certificates issued after an MFA check.
// See https://github.com/gravitational/teleport/issues/18544.
CertExtensionMFAVerifiedSessionExpires = "mfa-verified-session-expires"

// MFAVerifiedSessionExpiresASN1ExtensionOID is the RFC3339 timestamp representing the hard
// deadline of the session on a certificates issued after an MFA check.
// See https://github.com/gravitational/teleport/issues/18544.
MFAVerifiedSessionExpiresASN1ExtensionOID = asn1.ObjectIdentifier{1, 3, 9999, 2, 12}
```

These get set when short lived (<1m) single-use MFA certs are issued, and ultimately passed to the connection monitor as the `DisconnectExpiredCert` value, which fixes https://github.com/gravitational/teleport/issues/18544

### Manual testing

See instructions in [webapps README](https://github.com/gravitational/webapps/blob/1e133f2e4975dc91db6f2cf1c8a14dcd7ecdf770/README.md?plain=1#L253-L271) for local mfa development. My auth service looks like
```yaml
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  authentication:
    second_factor: optional #off
    require_session_mfa: yes
    webauthn:
      rp_id: proxy.127.0.0.1.nip.io
  disconnect_expired_cert: yes
```

To confirm the force-disconnection time on mfa sessions, I added the line

```go
fmt.Printf("Monitor force-disconnecting in %v\n", discTime)
```

below this one in `monitor.go`
https://github.com/gravitational/teleport/blob/b571efe990b1b2db089199f002f4d83075eee335/lib/srv/monitor.go#L181

Before these changes it would show ~56seconds, whereas now it shows a time closer to the cluster's default ttl (12h in my case).

**Behavior manually confirmed for**
- [x] SSH (@ibeckermayer)
- [x] Desktop Access (@ibeckermayer)
- [x] Kubernetes (@tigrato)
- [x] App Access (@mdwn): will be double checked once https://github.com/gravitational/teleport/pull/18607#issuecomment-1323967391 is addressed, we can proceed with merging this fix for now.
- [x] DB Access (@smallinsky)

### CODE REVIEW

File: lib/kube/proxy/forwarder.go [hunk 1]
```diff
@@ -691,8 +691,14 @@ func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUs
...
-	if !certExpires.IsZero() && disconnectExpiredCert {
-		authCtx.disconnectExpiredCert = certExpires
+	if disconnectExpiredCert {
+		if !identity.MFAVerifiedSessionExpires.IsZero() {
+			// Cause MFA verified sessions to disconnect on issuing certs expiry
+			// (see https://github.com/gravitational/teleport/issues/18544).
+			authCtx.disconnectExpiredCert = identity.MFAVerifiedSessionExpires
+		} else if !certExpires.IsZero() {
+			authCtx.disconnectExpiredCert = certExpires
+		}
```
Comment (ibeckermayer) [inline lib/kube/proxy/forwarder.go:701]: I need some guidance here. I suspect this logic is wrong however I know next to nothing about how kubernetes access works in general or what this section is doing in particular.

The typical process for setting `disconnectExpiredCert` is encoded in `GetDisconnectExpiredCertFromIdentity`
https://github.com/gravitational/teleport/blob/6263b81de3ea284c6b93746a7c56a0ae954496f2/lib/srv/monitor.go#L429-L457

However because we are using a special value `certExpires` here, rather than the typical case of using `identity.Expires`, I suspect that using `identity.MFAVerifiedSessionExpires` on line 698 is wrong. Probably I instead want to parse the cert that `certExpires` come from and use the new extension (see top level description) that corresponds to `MFAVerifiedSessionExpires`, but I need somebody who actually understands the system to confirm.

Tagging those who `git blame` tells me have worked on code around here: @dboslee @tigrato @Joerger @smallinsky @klizhentas

Comment (ibeckermayer) [reply lib/kube/proxy/forwarder.go]: Here's what I think it should actually look like, still please somebody double check that this makes sense in context:

919aca472825265ea285559e2ec2fe544c4564aa

File: api/proto/teleport/legacy/types/events/events.proto [hunk 2]
```diff
@@ -3211,6 +3211,13 @@ message Identity {
     (gogoproto.jsontag) = "allowed_resource_ids,omitempty",
     (gogoproto.nullable) = false
   ];
+  // MFAVerifiedSessionExpires is the the hard deadline of a session when this Identity was
+  // confirmed immediately after an MFA check.
+  google.protobuf.Timestamp MFAVerifiedSessionExpires = 23 [
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/events/events.proto:3216]: I think we can do a better job with the name. Remember, the root issue is that the "disconnect expired cert" option doesn't work with session MFA. This is not really a session expiry, because there are valid use cases where a cert expires but the session is allowed to continue.

How about

File: lib/auth/grpcserver.go [hunk 3]
```diff
@@ -2356,7 +2356,9 @@ func userSingleUseCertsAuthChallenge(gctx *grpcContext, stream proto.AuthService
 	return mfaDev, nil
 }
 
-func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice) (*proto.SingleUseUserCert, error) {
+// mfaVerifiedSessionExpires is the expiry of the user cert from which this single use cert is derived from.
+// See https://github.com/gravitational/teleport/issues/18544.
+func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice, mfaVerifiedSessionExpires time.Time) (*proto.SingleUseUserCert, error) {
```
Comment (zmb3) [inline lib/auth/grpcserver.go:2361]:
```suggestion
func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice) (*proto.SingleUseUserCert, error) {
```

You don't need to pass in the expiry as an argument, you already have it in `actx`.

File: lib/auth/grpcserver.go [hunk 4]
```diff
@@ -2356,7 +2356,9 @@ func userSingleUseCertsAuthChallenge(gctx *grpcContext, stream proto.AuthService
 	return mfaDev, nil
 }
 
-func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req proto.UserCertsRequest, mfaDev *types.MFADevice) (*proto.SingleUseUserCert, error) {
+// mfaVerifiedSessionExpires is the expiry of the user cert from which this single use cert is derived from.
```
Comment (zmb3) [inline lib/auth/grpcserver.go:2359]: Is this comment misplaced?

File: lib/auth/permissions.go [hunk 5]
```diff
@@ -297,18 +302,19 @@ func (a *authorizer) authorizeRemoteUser(ctx context.Context, u RemoteUser) (*Co
...
-		MFAVerified:       u.Identity.MFAVerified,
-		ClientIP:          u.Identity.ClientIP,
-		PrivateKeyPolicy:  u.Identity.PrivateKeyPolicy,
+		Usage:                     u.Identity.Usage,
+		RouteToCluster:            u.Identity.RouteToCluster,
+		KubernetesCluster:         u.Identity.KubernetesCluster,
+		RouteToApp:                u.Identity.RouteToApp,
+		RouteToDatabase:           u.Identity.RouteToDatabase,
+		MFAVerified:               u.Identity.MFAVerified,
+		MFAVerifiedSessionExpires: mfaVerifiedSessionExpires,
```
Comment (zmb3) [inline lib/auth/permissions.go:315]: This is either going to be empty, or identical to `Expires` right?

If that's the case, do we need to set it here or can we rely on the existing `Expires` and `MFAVerified` fields?

File: lib/services/authority.go [hunk 6]
```diff
@@ -287,6 +287,9 @@ type UserCertParams struct {
 	// MFAVerified is the UUID of an MFA device when this Identity was
 	// confirmed immediately after an MFA check.
 	MFAVerified string
+	// MFAVerified is the the hard deadline of a session when this Identity was
```
Comment (zmb3) [inline lib/services/authority.go:290]: Same comment as above. It's not a hard deadline, because if "disconnect expired cert" is disabled then it does absolutely nothing.

How about something like `OriginalIdentityExpires`?

File: lib/srv/monitor.go [hunk 7]
```diff
@@ -424,3 +425,61 @@ func (t *TrackingReadConn) UpdateClientActivity() {
...
+	// In the case where both disconnect_expired_cert and require_session_mfa are enabled,
+	// the MFAVerifiedSessionExpires value of the certificate will be used, which is the
+	// expiry of the certificate used to issue the short lived MFA verified certificate.
+	//
+	// See https://github.com/gravitational/teleport/issues/18544
+
+	// If the session doesn't need to be disconnected on cert expiry just return the default value.
+	if !checker.AdjustDisconnectExpiredCert(authPref.GetDisconnectExpiredCert()) {
+		var t time.Time
+		return t
```
Comment (zmb3) [inline lib/srv/monitor.go:446]:
```suggestion
		return time.Time{}
```

**Review by zmb3 — COMMENTED** [review]
Overall change looks good.

I think the godocs/naming can be improved, and we might be able to reduce some of the plumbing and set the original expiry at the last possible moment before we issue the MFA certs, but I have Friday afternoon brain so I'll take a fresh look again next week.

File: lib/kube/proxy/forwarder.go (hunk 1 above)
Comment (tigrato) [reply lib/kube/proxy/forwarder.go]: Kubernetes Proxy caches users' certificates and uses the `certExpires` (previously not after) to differentiate certificates for the same user (they will end up with different keys in the cache). This is important for cases where the user has connected to a proxy with a certificate that missed some Traits and has to re-login to regain them (check #18070 for more details)

As far as I can see, `Expires` is not marshaled into the certificate and it's not carried around. It depends on what is passed as the second arg to `tlsca.FromSubject` function.

```go
	id, err := tlsca.FromSubject(csr.Subject, time.Time{})
// or
        id, err := tlsca.FromSubject(csr.Subject,csr.NotAfter)
```
 and we have cases where we pass a zero time [[1](https://github.com/gravitational/teleport/blob/e17bcb13014e7b71bf787b3fcab5e8bd754d3b95/lib/auth/db.go#L152) ,[2](https://github.com/gravitational/teleport/blob/e17bcb13014e7b71bf787b3fcab5e8bd754d3b95/lib/auth/kube.go#L99)].

Since you pass the `clientCert.NotAfter` as a second arg to `tlsca.FromSubject`, `Expires` will be populated correctly.

When disconnect on cert expiration is enabled, the `disconnectExpiredCert` should be the cert expiration date if MFA is disabled or the `identity.MFAVerifiedSessionExpires` when MFA is enabled. 

From what I saw, your changes are correct and working accordingly.

File: api/proto/teleport/legacy/types/events/events.proto (hunk 2 above)
Comment (ibeckermayer) [reply api/proto/teleport/legacy/types/events/events.proto:3216]: (Looks like you forgot to include your suggestion)

Yeah good point. How about `IssuingCertExpires`?

Comment (ibeckermayer) [reply api/proto/teleport/legacy/types/events/events.proto]: 73889219589f717c33193070365f12b631ae64c3

File: lib/services/authority.go (hunk 6 above)
Comment (ibeckermayer) [reply lib/services/authority.go:290]: https://github.com/gravitational/teleport/pull/18607#discussion_r1028409180

File: lib/auth/grpcserver.go (hunk 4 above)
Comment (ibeckermayer) [reply lib/auth/grpcserver.go]: I think I was trying to document what the `previousIdentityExpires` parameter does, but that's documented enough elsewhere that I'll just get rid of it.

File: lib/auth/permissions.go (hunk 5 above)
Comment (ibeckermayer) [reply lib/auth/permissions.go]: @zmb3 I'm not sure what I was thinking, I think this makes sense: b571efe990b1b2db089199f002f4d83075eee335

@tigrato I'm not seeing how that relates

File: lib/kube/proxy/forwarder.go (hunk 1 above)
Comment (ibeckermayer) [reply lib/kube/proxy/forwarder.go]: Thanks for testing

File: lib/auth/permissions.go [hunk 8]
```diff
@@ -273,6 +273,12 @@ func (a *authorizer) authorizeRemoteUser(ctx context.Context, u RemoteUser) (*Co
 	// Adjust expiry based on locally mapped roles.
 	ttl := time.Until(u.Identity.Expires)
 	ttl = checker.AdjustSessionTTL(ttl)
+	var previousIdentityExpires time.Time
+	if u.Identity.MFAVerified != "" {
+		prevIdentityTTL := time.Until(u.Identity.Expires)
```
Comment (tigrato) [inline lib/auth/permissions.go:278]: Isn't this the same as ttl?

**Comment by mdwn** [comment]
So I'm not sure that app access actually works with MFA enabled at the moment. I'm trying to figure out why. That being said, maybe not worth holding up on that verification.

File: lib/auth/permissions.go (hunk 8 above)
Comment (ibeckermayer) [reply lib/auth/permissions.go]: Yes, it was, changed to what I meant it to be: a26f201546be180648fceee39c15da2abe427c49

File: lib/auth/permissions.go (hunk 5 above)
Comment (ibeckermayer) [reply lib/auth/permissions.go]: > From what I understood, @zmb3 was asking to remove PreviousIdentityExpires here since its value is either empty or equal to Expires, and re-use the value of Expires and MFAVerified to decide the PreviousIdentityExpires.

I think what was meant was: at the time @zmb3's original comment was written, [the logic for setting `MFAVerifiedSessionExpires`](https://github.com/gravitational/teleport/pull/18607/files/919aca472825265ea285559e2ec2fe544c4564aa#diff-a0d9859c8f6c722d6937c51e77f3db91423e15cfebc385111d891ef30cdf218fR292-R296) (now called `PreviousIdentityExpires`) meant that it would either be empty or just the value of `Expires`, so it didn't really need to be set here.

However I don't know what I was thinking when I wrote that. My understanding is that what we're doing here is converting a local cluster identity to an identity that original identity is allowed to take up on a remote cluster. 

What I'm doing between https://github.com/gravitational/teleport/commit/b571efe990b1b2db089199f002f4d83075eee335 + https://github.com/gravitational/teleport/commit/a26f201546be180648fceee39c15da2abe427c49 is copying the original local cluster identity's `PreviousIdentityExpires` trait over, while ensuring it doesn't exceed the max allowed TTL for this new remote cluster identity by passing it through `AdjustSessionTTL` (the same process that was already being done for `Expires`).

File: lib/services/authority.go (hunk 6 above)
Comment (tigrato) [inline lib/services/authority.go:290]: Comment typo

File: lib/srv/monitor.go [hunk 9]
```diff
@@ -424,3 +426,60 @@ func (t *TrackingReadConn) UpdateClientActivity() {
...
+	// expiry of the certificate used to issue the short lived MFA verified certificate.
+	//
+	// See https://github.com/gravitational/teleport/issues/18544
+
+	// If the session doesn't need to be disconnected on cert expiry just return the default value.
+	if !checker.AdjustDisconnectExpiredCert(authPref.GetDisconnectExpiredCert()) {
+		return time.Time{}
+	}
+
+	if identity.MFAVerified != "" && !identity.PreviousIdentityExpires.IsZero() {
```
Comment (tigrato) [inline lib/srv/monitor.go:449]: Since PreviousIdentityExpires is only set when MFA is required, can't we just use 

```suggestion
	if !identity.PreviousIdentityExpires.IsZero() {
```

File: lib/srv/monitor.go [hunk 10]
```diff
@@ -424,3 +426,60 @@ func (t *TrackingReadConn) UpdateClientActivity() {
...
+	authPref types.AuthPreference,
+	identity *IdentityContext,
+) time.Time {
+	// In the case where both disconnect_expired_cert and require_session_mfa are enabled,
+	// the PreviousIdentityExpires value of the certificate will be used, which is the
+	// expiry of the certificate used to issue the short lived MFA verified certificate.
+	//
+	// See https://github.com/gravitational/teleport/issues/18544
+
+	// If the session doesn't need to be disconnected on cert expiry just return the default value.
```
Comment (tigrato) [inline lib/srv/monitor.go:471]: it's probably better to reuse the same code in both functions

**Comment by ibeckermayer** [comment]
Noticed we didn't have any sanity check test of this kind so I added it here https://github.com/gravitational/teleport/commit/faeed096b10553c4c7953388948cdbdb63507dd2#diff-3057c3148b41a175f31650e1b9033e2148353a0d4b43b79404d0fdfe0330f56aR176-R198

File: constants.go [hunk 11]
```diff
@@ -439,6 +439,12 @@ const (
 	// CertExtensionMFAVerified is used to mark certificates issued after an MFA
 	// check.
 	CertExtensionMFAVerified = "mfa-verified"
+	// CertExtensionPreviousIdentityExpires is the RFC3339 timestamp representing the
+	// expiry time of the identity/cert that this identity/cert was derived from. It
+	// is used to determine a session's hard deadline in cases where both
+	// require_session_mfa and disconnect_expired_cert are enabled.
+	// See https://github.com/gravitational/teleport/issues/18544.
+	CertExtensionPreviousIdentityExpires = "prev-identity-expires"
```
Comment (zmb3) [inline constants.go:447]:
```suggestion
	// CertExtensionPreviousIdentityExpires is an extension that stores the RFC3339 timestamp representing the
	// expiry time of the identity/cert that this identity/cert was derived from. It
	// is used to determine a session's hard deadline in cases where both
	// require_session_mfa and disconnect_expired_cert are enabled.
	// See https://github.com/gravitational/teleport/issues/18544.
	CertExtensionPreviousIdentityExpires = "prev-identity-expires"
```

This variable itself is not a timestamp, so I would reword things a bit.

File: api/proto/teleport/legacy/types/events/events.proto [hunk 12]
```diff
@@ -3211,6 +3211,15 @@ message Identity {
     (gogoproto.jsontag) = "allowed_resource_ids,omitempty",
     (gogoproto.nullable) = false
   ];
+  // PreviousIdentityExpires is the expiry time of the identity/cert that this
+  // identity/cert was derived from. It is used to determine a session's hard
```
Comment (zmb3) [inline api/proto/teleport/legacy/types/events/events.proto:3215]: Again, I wouldn't use the `hard deadline` verbiage, since the default mode is for disconnect on expired cert to be false (which means it isn't actually a "hard deadline" unless you opt-in to that behavior)

File: lib/auth/grpcserver.go [hunk 13]
```diff
@@ -2372,7 +2372,8 @@ func userSingleUseCertsGenerate(ctx context.Context, actx *grpcContext, req prot
 	}
 
 	// Generate the cert.
-	certs, err := actx.generateUserCerts(ctx, req, certRequestMFAVerified(mfaDev.Id), certRequestClientIP(clientIP))
+	previousIdentityExpires := actx.Identity.GetIdentity().Expires
+	certs, err := actx.generateUserCerts(ctx, req, certRequestMFAVerified(mfaDev.Id), certRequestPreviousIdentityExpires(previousIdentityExpires), certRequestClientIP(clientIP))
```
Comment (zmb3) [inline lib/auth/grpcserver.go:2376]: This is long enough I'd probably wrap each of the `certRequestXYZ` options on its own line. And then inline the `previousIdentityExpires` variable.

File: api/proto/teleport/legacy/types/events/events.proto (hunk 12 above)
Comment (ibeckermayer) [reply api/proto/teleport/legacy/types/events/events.proto:3215]: I think it's accurate now, it says "It is used to determine a session's hard deadline **in cases where both require_session_mfa and disconnect_expired_cert are enabled**."

File: lib/auth/test/suite.go [hunk 14]
```diff
@@ -172,6 +172,30 @@ func (s *AuthSuite) GenerateUserCert(t *testing.T) {
...
+		CertificateFormat:       constants.CertificateFormatStandard,
+		MFAVerified:             "mfa-device-id",
+		PreviousIdentityExpires: clock.Now().Add(time.Hour),
+	})
+	require.NoError(t, err)
+	parsedCert, err = sshutils.ParseCertificate(cert)
+	require.NoError(t, err)
+	devId, ok := parsedCert.Extensions[teleport.CertExtensionMFAVerified]
+	require.True(t, ok)
+	require.Equal(t, "mfa-device-id", devId)
```
Comment (zmb3) [inline lib/auth/test/suite.go:193]: Up to you, but:
```suggestion
	require.Contains(t, parsedCert.Extensions, teleport.CertExtensionMFAVerified)
	require.Equal(t, "mfa-device-id", parsedCert.Extensions[teleport.CertExtensionMFAVerified])
```

File: lib/kube/proxy/forwarder.go [hunk 15]
```diff
@@ -536,7 +541,7 @@ func (f *Forwarder) formatResponseError(rw http.ResponseWriter, respErr error) {
 	}
 }
 
-func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUser bool, certExpires time.Time) (*authContext, error) {
+func (f *Forwarder) setupContext(ctx auth.Context, req *http.Request, isRemoteUser bool, clientIdentity *tlsca.Identity) (*authContext, error) {
```
Comment (zmb3) [inline lib/kube/proxy/forwarder.go:544]: While here, can we use another var name for the `auth.Context`? I prefer to keep `ctx` reserved for `context.Context`.

File: lib/srv/monitor.go [hunk 16]
```diff
@@ -424,3 +425,59 @@ func (t *TrackingReadConn) UpdateClientActivity() {
 	defer t.mtx.Unlock()
 	t.lastActive = t.cfg.Clock.Now().UTC()
 }
+
+// GetDisconnectExpiredCertFromIdentity calculates the proper value for DisconnectExpiredCert
+// based on whether a connection is set to disconnect on cert expiry, and whether
+// the cert is a short lived (<1m) one issued for an MFA verified session.
```
Comment (zmb3) [inline lib/srv/monitor.go:431]: I'd mention under what circumstances it returns a zero time.

**Review by zmb3 — APPROVED** [review]

**Review by tigrato — APPROVED** [review]

**Comment by ibeckermayer** [comment]
@smallinsky @Joerger since I'm not familiar with this part of the code base, I'd like one of you guys to confirm that something for me, which boils down to confirming that we were never relying on calling `func (f *Forwarder) authenticate(req *http.Request) (*authContext, error)` on a request with a cert that was missing either of the fields I added (to a test) in a07e91315f561e1463972631ddf1b78d37abd5af.

The broader context is:

Initially my changes managed to break one of the forwarder tests (https://github.com/gravitational/teleport/issues/18928). I tracked it down, and the reason was that I'd added [this line](https://github.com/gravitational/teleport/pull/18607/files#diff-cdfe0b66f7090f9a5e0759d6cce323f7dde1ee1f0f49402d6a7652cd91b69430R436) to `func (f *Forwarder) authenticate(req *http.Request) (*authContext, error)`, which calls 

https://github.com/gravitational/teleport/blob/a07e91315f561e1463972631ddf1b78d37abd5af/lib/tlsca/ca.go#L292-L302 which needs a `Username` and `Groups`. I fixed it by modifying the test like this: a07e91315f561e1463972631ddf1b78d37abd5af.

**Comment by smallinsky** [comment]
@ibeckermayer Regarding: 
https://github.com/gravitational/teleport/pull/18607#issuecomment-1339938629

Based on my limited knowledge about teleport k8 access adding the [tlsca.FromSubject](https://github.com/gravitational/teleport/pull/18607/files#diff-cdfe0b66f7090f9a5e0759d6cce323f7dde1ee1f0f49402d6a7652cd91b69430R436) seems to be reasonable. 

The `(f *Forwarder) authenticate` function works only for `auth.RemoteUser` and `auth.LocalUser` contexts 
where  `tlsca.FromSubject`  is already called to create teleport context by the HTTP middleware: 
https://github.com/gravitational/teleport/blob/8c1e7be02dfec643d9796eff7a90a2119ac7e891/lib/auth/middleware.go#L489

**Comment by ibeckermayer** [comment]
v9 has diverged substantially from v11 in various ways that make backporting difficult. I'm going to just leave this as-is for now, and can preform the surgical v9 backport if that's required by a particular customer.
